### PR TITLE
Fix #55

### LINKED ISSUE #55: BCnearest bug
Here's an minimal example. Sorry for the format:

clipboard-1

1d linear interpolation + BCnearest extrapolation. Values higher than the grid look fine but values below the bottom value of the grid are all wrong.

Any suggestions?

**Edit:** Version 0.3.2 (2014-10-21 22:05 UTC)

### PR DESCRIPTION
**Pull request by tlycken**
Thanks to @mbauman's troubleshooting it seems this was a quite easy fix; the following plots look reasonable for both linear and quadratic interpolation (only linear shown):

```julia
# using Gadfly, DataFrames
reload("Grid")

f(x) = sin(2pi * (x-1) / 10)
dfs = Array(DataFrame,0)

for (i,bc) in enumerate([
        Grid.BCnan,
        Grid.BCna,
        Grid.BCperiodic,
        Grid.BCreflect,
        Grid.BCnearest,
        Grid.BCfill])
    itp = bc != Grid.BCfill ? Grid.InterpGrid(f(1:11), bc, Grid.InterpLinear) :
        Grid.InterpGrid(f(1:11), 1, Grid.InterpLinear)
    push!(dfs, DataFrame(x=-9:.1:40,y=[itp[x]+(i-1)*3 for x in -9:.1:40],t="$bc"))
end

df = vcat(dfs)
plot(df,x=:x,y=:y,color=:t,Geom.path,Guide.colorkey("Boundary condition"))
```

bcs

I didn't test cubic interpolation yet simply because I know that it's not rock solid anyway, so I wanted to get this up for a Travis run first. (Did any tests break from this behavior before? If not, what's the best way to test that this actually works the way we want it?)

### CODE REVIEW

**Comment by tlycken** [comment]
Failed on 0.4 because of Woodbury matrices - will merging #56 fix that?

I think a 20% coverage decrease because of these changes is very unlikely - could the difference be due to the "new" metric? (Has coverage results not been collected since it was introduced?)

**Comment by mbauman** [comment]
YAY!  Nice work.  Is that one-sample hold in BCreflect and BCperiodic new?  Or is that an existing issue?  Or is it expected behavior?  (I've never really used boundary conditions)

**Comment by tlycken** [comment]
@mbauman The one-cell shift is actually the desired behavior in some applications (notably for images, which happens to be the application Tim originally built the library for) - basically, it is to make sure that `x[n+1:2n] == reverse(x)`. So it's not really a bug *per se*, although it's not the expected behavior in many other applications (for example when extrapolating `sin(x)`, which happens to be the example I used...). Interpolations.jl has taken this into account by design, and so lets the user choose if they want this behavior or the one you were expecting.

tl;dr: it's neither new, unknown, optimal or unexpected :)

**Comment by mbauman** [comment]
:+1:

**Comment by mbauman** [comment]
I should have pushed that WIP a long time ago.  Thanks for finishing it up.

**Comment by tlycken** [comment]
np. I had to change a total of one extra line :)

**Comment by timholy** [comment]
Thanks, folks. I am just swamped, and it's wonderful to have such awesome collaborators.

**Comment by tlycken** [comment]
:cocktail: :tada: :beers:

**Comment by timholy** [comment]
Re the coverage stats: it's likely that the test failure caused the decrease. When it fails a tests, it doesn't run as much code, but it still submits the results to Coveralls.

**Comment by tlycken** [comment]
The build still fails on 0.4 with the message `Woodbury is not defined`; it turns out that it doesn't load the package, since [the ppa for nightlies is still on an earlier build](https://launchpad.net/~staticfloat/+archive/ubuntu/julianightlies), but apparently it isn't available anyway. I'll drill down and see if I can figure out why.

**Comment by timholy** [comment]
The "push" failed, but the "pr" (pull request) succeeded. I'm not certain of the difference between them, but if I had to guess I'd say the "push" adds on to the branch that you started with, whereas the "pr" is for what would happen should you merge this to master. Since your Woodbury fix occurred after you got this started, I think that explains the difference.

So I'd say, merge at will.

**Comment by tlycken** [comment]
Yeah, that might explain it. Might as well rebase to test the theory :)
